### ISSUE TITLE
Empty error: Is there a debug of verbose option?

### ISSUE THREAD

**Issue by nikp** [simonw/llm#189]
Hi there, I'm just getting started with this tool (fantastic abstractions, kudos!), using this guide: https://simonwillison.net/2023/Aug/1/llama-2-mac/

Unfortunately, while all the prerequisites succeeded without any issues, running the actual prompt results in an empty error. Is there any way to run it to reveal more debug information?

```
➜  ~/llm➜ llm -m l2c 'Tell me a joke about a llama'

Error: 
```

Some context if it would be helpful:
```
➜  ~/llm➜ which llm
/usr/local/bin/llm
➜  ~/llm➜ python --version
Python 3.11.5
➜  ~/llm➜ stat /Users/nppinski/Library/Application\ Support/io.datasette.llm/llama-cpp/models/llama-2-7b-chat.ggmlv3.q8_0.bin
16777230 157262586 -rw-r--r-- 1 nppinski staff 0 7160799872 "Aug 27 15:55:00 2023" "Aug 27 15:55:00 2023" "Aug 27 15:55:00 2023" "Aug 27 15:46:31 2023" 4096 13993408 0 /Users/nppinski/Library/Application Support/io.datasette.llm/llama-cpp/models/llama-2-7b-chat.ggmlv3.q8_0.bin
```

**Comment by af** [comment]
Ran into the same issue, FWIW I installed via `brew install llm`

On macOS 13.4.1 with Python 3.11.5

**Comment by af** [comment]
@nikp Still not sure what my issue was, but I tried this alternate setup step from the link in this issue and it now works for me 🎉 :

```
llm install https://static.simonwillison.net/static/2023/llama_cpp_python-0.1.77-cp311-cp311-macosx_13_0_arm64.whl
```

**Comment by nikp** [comment]
OK, so this is what's strange @af, I did try running that but got 

```
ERROR: llama_cpp_python-0.1.77-cp311-cp311-macosx_13_0_arm64.whl is not a supported wheel on this platform.
```

Same mac, same version of Python, but I am on OS X `13.5.1`. I wondered if that would make a difference.

**Comment by af** [comment]
I actually upgraded to 13.5.1 before getting things running finally, so our OS version was even the same 🤷

**Comment by jokoso** [comment]
I'm using the guide as well and get the same error when using `llama-2` models from HF. `orca-mini` and `wizard` models from `gpt4all` work fine. It seems that anything using `llama-ccp` fails with the empty error:

```
Error: 
```

I'm on Linux and Python 3.11.

**Comment by band** [comment]
I reinstalled everything this morning and I have the same empty error as @jokoso when trying to use `llama-2-7b-chat.ggmlv3.q8_0`. other models work fine.

I'm on macOS 13.5.1 and Python 3.11.5

**Comment by phase** [comment]
> @nikp Still not sure what my issue was, but I tried this alternate setup step from the link in this issue and it now works for me 🎉 :
> 
> ```
> llm install https://static.simonwillison.net/static/2023/llama_cpp_python-0.1.77-cp311-cp311-macosx_13_0_arm64.whl
> ```

This worked for me!

**Comment by dannguyen** [comment]
Am chiming in from Windows/WSL Ubuntu 20.04.6 land, to add that I am also getting the empty `Error: ` when trying to do `llm -m l2c 'hello'`. I followed all the steps in simonw's blog with the exception of `brew install llm`.

The ChatGPT-usage of llm works fine. I'll experiment with other models but if there's a way to get verbose debug output from llm, I can plunk around with `-m l2c` specifically.

**Comment by aaron13100** [comment]
I also get this empty error message on Ubuntu following the doc (https://github.com/simonw/llm-llama-cpp#adding-models) whether the file is downloaded already or not.

**Comment by jjohnson1024** [comment]
This error appears to be related to the llama-cpp-python version. At the time I'm writing this, the latest version is 0.1.84. Installing the 0.1.77 version (matching the version of the pre-compiled wheel in the guide) clears up the issue for me.

`llm install --force-reinstall --no-cache-dir llama-cpp-python==0.1.77`

**Comment by vividfog** [comment]
**TLDR: You can `pip install` the latest `llama-cpp-python` with no flags, IF you drop a missing `ggml-metal.metal` file into the `llama_cpp` `site-packages` afterwards.**

Step by step:

The latest version can be made to work with Metal, but it requires manually moving a file into the correct directory. The problem is related to `llama-cpp-python`, tangential to what people are talking about here: https://github.com/abetlen/llama-cpp-python/issues/509

The final fix is in the hands of the dependency, so that `pip install llama-cpp-python` is enough to install the Metal support files in a way that works. They have instructions for a Metal build, but those didn't work for me. Build errors of all kinds.

### In the meanwhile, this seems to work:

1. `pip install llama-cpp-python` (today it was version v0.1.84)

2. Clone ggerganov's original `llama.cpp` repository and follow the Metal instructions for Release to build the missing file: https://github.com/ggerganov/llama.cpp

At this point it might be wise to test your newly built ./main binary to ensure it works. Follow the instructions in the repository. Check that it uses your GPU, it streams and it doesn't truncate the answers.

`./main -m local_model_file.gguf -p "A long essay on why orcas don't like captivity: \n" -n 1024 -c 1024 -e -ngl 1`

Local models are finicky and often require special prompts. Try with many before concluding that something is already wrong at this step. This model worked for me: https://huggingface.co/TheBloke/orca_mini_v3_7B-GGUF/resolve/main/orca_mini_v3_7b.Q4_K_M.gguf

(And looking at that command, that is why we'd rather prefer to use `llm`, isn't it...)

### If so far so good:

3. Inside the `build/bin` folder you fill find a file called `ggml-metal.metal` ... and if `llama.cpp` can't find this at runtime, there will be an `Error:`.
4. The correct place for this file is in your environment `site_packages`, inside the `llama_cpp` folder. For me this happens to be `/opt/homebrew/Caskroom/miniconda/base/lib/python3.11/site-packages/llama_cpp` but for you it will be different depending on how your Python is installed. Just drop `ggml-metal.metal` there and Metal should now work, `Error:` is gone.

If you're interested in seeing actual error message before this operation, add `-o verbose true` after the `llm -m llama-based-model.gguf` command. That's the verbose mode someone asked.

Before:

```
ggml_metal_init: error: Error Domain=NSCocoaErrorDomain Code=258 "The file name is invalid."
llama_new_context_with_model: ggml_metal_init() failed
```

After:

```
> llm -m orca_mini_v3_7b.Q4_K_M "A long essay on why orcas don't like captivity:"
Orcas are highly intelligent, social, and adaptable marine mammals that belong to the dolphin family. Bottlenose dolphins, another type of dolphin, are well known for their intelligence, social behavior, and playful nature. Although orcas and bottlenose dolphins have some similarities in appearance, they have vastly different personalities and behaviors. Orcas generally live... [it keeps going]
```

Supposedly this can also be made to work with an environment variable, but I didn't try that route. Note that if you later update your llama-cpp-python library (or some other project updates it for you), you may need to repeat this operation, unless the fix already included in the new version.

### Truncated output, no streaming?

If after this you still find that the outputs are truncated or the output is not streaming, there's a few-liner fix for both by patching `llm_llama_cpp.py`. Maybe @simonw will include these in a future version of this plugin.

Until then, git clone this plugin and follow these:

https://github.com/simonw/llm-llama-cpp/issues/6#issuecomment-1664960492
https://github.com/simonw/llm-llama-cpp/issues/11#issuecomment-1712900506

Then `pip install -e .` while in the plugin directory with the patched plugin.

Tested on M1 Max. Truncation/streaming patch was also tested on a Linux box with no GPU.

**Comment by nikp** [comment]
> This error appears to be related to the llama-cpp-python version. At the time I'm writing this, the latest version is 0.1.84. Installing the 0.1.77 version (matching the version of the pre-compiled wheel in the guide) clears up the issue for me.
> 
> `llm install --force-reinstall --no-cache-dir llama-cpp-python==0.1.77`
> 
> Edit: This issue seems to get at the root of the problem: [simonw/llm-llama-cpp#10](https://github.com/simonw/llm-llama-cpp/issues/10)
> 
> llama-cpp moved from using GGML files to GGUF files, so an alternative is to patch llm-lamma-cpp so that GGUF files can be downloaded (details all in the linked issue).

@jjohnson1024, thank you! Force reinstalling this dependency fixed things for me..

**Comment by vividfog** [comment]
For more output from llm-llama-cpp plugin, add `-o verbose true` or look at all model options with `llm models --options`

`llm -m llama-2-7b-chat.Q8_0 "Write lyrics for a song about sea" -o verbose true`

There's a way to use the latest llama-cpp-python and llama.cpp with GGUF support, but the process is currently a bit hands-on. If you're OK with that: https://github.com/simonw/llm-llama-cpp/issues/14

**Comment by aaron13100** [comment]
> llama-cpp moved from using GGML files to GGUF files, so an alternative is to patch llm-lamma-cpp so that GGUF files can be downloaded (details all in the linked issue).

Based on this and the error message when using "-o verbose true" that said it couldn't load the gguf file (although I had specified a ggml file), I was able to get this to work by downloading a gguf file and loading that instead. I used llama-2-70b-chat.Q8_0.gguf from https://huggingface.co/TheBloke/Llama-2-70B-chat-GGUF

**Comment by ziyunli** [comment]
From https://huggingface.co/TheBloke/Llama-2-13B-chat-GGML/discussions/14 

> latest llama.cpp is no longer compatible with GGML models.

The thread mentions you can convert the model from GGML format to GGUF, but if you're following Simon's tutorial, you can just download the GGUF model instead with something like:

```sh
llm llama-cpp download-model \
  https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGML/resolve/main/llama-2-7b-chat.ggmlv3.q8_0.bin \
  --alias llama2-chat --alias l2c --llama2-chat
```
